### PR TITLE
Make HFEM lower-D mesh work with distributed mesh

### LINKED ISSUE #17484: HFEM in debug mode with distributed mesh will hit assertions
## Reason
These assertions seems related with lower-d element neighbors, which is not needed by HFEM. Even we have these assertions, HFEM can run fine and deliver the correct solutions. But it is still nice to fix them.

## Design
These are some comments from @roystgnr :

>At the libMesh level... disabling that assertion would fix the immediate problem (since we never actually use the >nconsistent neighbor pointers on side elements) but would weaken our testing of a zillion other problems (since I can't >figure out a way to conditionally disable the assertion) and would potentially just cause us to fail other assertions as >soon as we tried repartitioning or adaptive refinement or anything else that can trigger a double-check of neighbor >pointer consistency.
>
>At the MOOSE level - we could try skipping prepare_for_use() entirely (calling the bits we still may want, like >renumber_nodes_and_elements() or update_parallel_id_counts(), and cache_elem_dims(), and well there's a lot... >manually), or we could try setting allow_find_neighbors(false) first. We aren't changing any neighbors on the high-d >element mesh, we don't need any neighbor links on the low-d element mesh, and the failing assertion just cares about >whether those links are consistent, not about whether they're complete.

## Impact
Make users able to run HFEM in debug mode with distributed mesh for code development.

### PR DESCRIPTION
**Pull request by lindsayad**
We achieve this by not initially allowing remote element removal
when doing HFEM because the HFEM mesh construction process currently
relies on the mesh being in a serialized state. But then eventually
we can remove all the remote elements.

Closes #17484

### CODE REVIEW

**Comment by YaqiWang** [comment]
Care to fix the line in `FEProblemBase` pointed by @yjung-anl during our conversation on Slack? Also with this change we can use distributed mesh for HFEM?

**Comment by lindsayad** [comment]
> Also with this change we can use distributed mesh for HFEM?

Your current logic in `buildLowerDMesh` relies on the incoming mesh being in a serial state; I haven't done anything to fix that. However, since your logic doesn't rely on the mesh being *replicated* I removed any logic that relied on `is_replicated`. Consequently, you can now run HFEM with `--distributed-mesh`, but we will not remove remote elements until right before the `init_problem` task.

**Comment by lindsayad** [comment]
> Care to fix the line in `FEProblemBase` pointed by @yjung-anl during our conversation on Slack

Done!

**Comment by YaqiWang** [comment]
> Consequently, you can now run HFEM with `--distributed-mesh`, but we will not remove remote elements until right before the `init_problem` task.

I thought we remove remote elements at setup_mesh_complete?

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -2141,7 +2141,7 @@ FEProblemBase::reinitElemNeighborAndLowerD(const Elem * elem, unsigned int side,
       std::vector<Point> reference_points;
       FEInterface::inverse_map(
           lower_d_elem_neighbor->dim(), FEType(), lower_d_elem_neighbor, qps, reference_points);
-      reinitLowerDElem(lower_d_elem_neighbor, tid, &qps);
+      reinitLowerDElem(lower_d_elem_neighbor, tid, &reference_points);
```
Comment (roystgnr) [inline framework/src/problems/FEProblemBase.C:2144]: Whoa.  How did nobody catch this bug before?

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -215,6 +215,10 @@ MooseMesh::MooseMesh(const InputParameters & parameters)
   else if (isParamValid("block"))
     _provided_coord_blocks = getParam<std::vector<SubdomainName>>("block");
 
+  if (getParam<bool>("build_all_side_lowerd_mesh"))
+    // Do not initially allow removal of remote elements
+    allowRemoteElementRemoval(false);
```
Comment (roystgnr) [inline framework/src/mesh/MooseMesh.C:220]: Where are we then restoring `allowRemoteElementRemoval(true)`?

Comment (lindsayad) [reply framework/src/mesh/MooseMesh.C:220]: Currently nowhere. But this method marks `_need_delete = true` if someone calls with `allowRemoteElementRemoval(false)`, and that flag is checked during the `delete_remote_elements_after_late_geometric_ghosting` MOOSE action task.

File: framework/src/problems/FEProblemBase.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:2144]: I don't know. I suspect the coverage report is going to tell me this line isn't hit. I suspect that we currently only ever call this method with the element that has the lower-D element associated with it, e.g. for non-refined meshes we are calling with the lower ID higher dimensional element etc.

**Review by roystgnr — APPROVED** [review]

**Pull request merged**